[SimplifyIndVar] Push more users to worklist for simplifyUsers

v01dXYZ opened the pull request:
Instead of considering only users that are inside the loop, consider all the users with parent dominated by the loop header.

This is a draft PR to discuss about the code  before doing some QA. @etherzhhb @efriedma-quic Do you think my code is valid ?

Provides a solution to #90417.

efriedma-quic commented:
Two potential problems with this approach:

- Some places might implicitly assume IV users are inside the loop.
- I'm a bit concerned about compile-time.  If you recursively visit every value derived from the exit value of the loop, you can easily end up with quadratic compile-time growth: if you have N successive loops, you visit instructions in the last loop N^2 times.

CC @fhahn

v01dXYZ commented:
The following breaks:

```
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"

define void @foo(
       i32 noundef %n_real_zones,
        ptr noundef %real_zones
) {
entry:
  br label %for.cond

for.cond:                                         ; preds = %for.cond.cleanup3, %entry
  %j.0 = phi i32 [ 0, %entry ], [ %inc7, %for.cond.cleanup3 ]
  %n_real_zones.addr.0 = phi i32 [ %n_real_zones, %entry ], [ %n_real_zones.addr.1, %for.cond.cleanup3 ]
  %cmp = icmp ult i32 %j.0, 128
  br label %for.cond1

for.cond1:                                        ; preds = %for.body4, %for.cond
  %n_real_zones.addr.1 = phi i32 [ %n_real_zones.addr.0, %for.cond ], [ %inc, %for.body4 ]
  %i.0 = phi i32 [ 0, %for.cond ], [ %inc5, %for.body4 ]
  %cmp2 = icmp ult i32 %i.0, 128
  br i1 %cmp2, label %for.body4, label %for.cond.cleanup3

for.cond.cleanup3:                                ; preds = %for.cond1
  %inc7 = add nuw nsw i32 %j.0, 1
  br label %for.cond

for.body4:                                        ; preds = %for.cond1
  %idxprom = sext i32 %n_real_zones.addr.1 to i64
  %arrayidx = getelementptr inbounds i32, ptr %real_zones, i64 %idxprom
  store i32 %j.0, ptr %arrayidx, align 4
  %inc = add nsw i32 %n_real_zones.addr.1, 1
  %inc5 = add nuw nsw i32 %i.0, 1
  br label %for.cond1
}
```

I don't know why it seems to depend on `datalayout`.

v01dXYZ commented:
I was able to pass the test-suite with the new changes.

etherzhhb commented:
Hi, i think this code assume we run the loop close SSA pass first. Could you include a LIT test to demonstrate the problem? I am not sure this is the right fix to your problem.

etherzhhb commented on llvm/lib/Transforms/Utils/SimplifyIndVar.cpp:
```diff
@@ -511,6 +511,7 @@ bool SimplifyIndvar::eliminateTrunc(TruncInst *TI) {
     ICmpInst *ICI = dyn_cast<ICmpInst>(U);
     if (!ICI) return false;
     assert(L->contains(ICI->getParent()) && "LCSSA form broken?");
+    //    assert(L->contains(ICI->getParent()) && "LCSSA form broken?");
```
Hi, i think this code assume we run the loop close SSA pass first. Could you include a LIT test to demonstrate the problem? I am not sure this is the right fix to your problem.

v01dXYZ replied:
This check will break if there is an exit (or dominated) block with a use-chain `trunc-icmp` such as the following one from `Bitcode/Regression/fft/x86_17.06.19_halide_runtime.bc halide_memoization_cache_lookup`: 

``` llvm
for.cond82.preheader:                             ; preds = %_ZNK15halide_buffer_t13size_in_bytesEv.exit
  %48 = trunc i64 %indvars.iv214 to i32
  %cmp83195 = icmp sgt i32 %48, 0
  br i1 %cmp83195, label %for.body85.preheader, label %cleanup111
```

or such as:

``` llvm
define i1 @foo(i32 %n) {
entry:
    br label %loop

loop:
    %i = phi i64 [0, %entry], [%i.inc, %loop]
    %i.inc = add i64 %i, 1

    %i.ult.Icnt = icmp ult i64 %i, 128

    br i1 %i.ult.Icnt, label %loop, label %loop.exit

loop.exit:
    %i.32 = trunc i64 %i to i32
    %i.32.ult.n = icmp ult i32 %i.32, %n
    ret i1 %i.32.ult.n
}
```

I've got to take a closer look at the code to understand why LCSSA is required here.

v01dXYZ replied:
It seems to me the utility class `SimplifyIndVar` don't have to be fed LCSSA-normalised loop but the loop pass `IndVarSimplify` needs it (I don't know exactly all the reasons, just spotted it uses `llvm::rewriteExitValue` which simply uses the Backedge Info + SCEV to replace the LCSSA phis with SCEV Constant if the phi is a SCEV Add Rec)

Maybe I'm making a mistake but `eliminateTrunc` checks all the `trunc` users are `icmp` that are eligible to simplifications. I don't see why the LCSSA is required here (except to ensure staying inside the loop).

v01dXYZ commented:
So I added a cl opt to limit the traversal *block-wise*. I'm trying to benchmark to see if it hurts performance.

v01dXYZ commented:
@efriedma-quic The simplification algorithm is almost greedy forward. It stops descending as soon as there is no possible simplification. There is one big exception with instructions representing a SCEV Add Rec associated to the current analysed loop which are *always* followed, but it's restricting and the user will be traversed only during the analysis of the right loop.

The quadratic behaviour could in theory appear when a simplification is always available. One way to avoid going too far is to keep a counter when traversing the use chain which is incremented everytime we jump to a out-of-loop block and reset when we go back inside the loop. It's a fast way to compute a overapproximation of `dist(Loop, CurrentBlock)`.

I ran `test-suite` for `compile_time` but I don't see significant regressions. I would like to have some advice about performance regressions as I fear I could make a dumb mistake (a short sanity check list).

fhahn commented:
FWIW it looks like we can handle the motivating case from https://github.com/llvm/llvm-project/issues/90417 in constraint-elimination: https://github.com/llvm/llvm-project/pull/94610

v01dXYZ commented:
Ping.

nikic commented:
If you want a review, move the PR out of draft.

v01dXYZ commented:
ping

nikic commented on llvm/include/llvm/Transforms/Utils/SimplifyIndVar.h:
```diff
@@ -52,12 +52,11 @@ class IVVisitor {
+simplifyUsersOfIV(PHINode *CurrIV, ScalarEvolution *SE, DominatorTree *DT,
+                  LoopInfo *LI, const TargetTransformInfo *TTI,
+                  SmallVectorImpl<WeakTrackingVH> &Dead, SCEVExpander &Rewriter,
+                  unsigned MaxDepthOutOfLoop = 1, IVVisitor *V = nullptr);
```
Do we need the MaxDepthOutOfLoop parameter? Can you move the cl::opt into SimplifyIndVar and directly use it instead?

v01dXYZ replied:
`Transforms/Utils/LoopUnroll.cpp` uses it via `simplifyLoopIVs`. I didn't want to add it here as it could create regressions in passes I'm not interested in.

nikic replied:
Okay, that makes sense. But as implemented, aren't you still calling this with the default of 1 from LoopUnroll? Shouldn't the default be 0 instead?

Could you please rebase over https://github.com/llvm/llvm-project/commit/6e3725d7f290d0180e8b5cfe073d0cc4c9bbafd0 so we at least don't have to pass through this parameter everywhere?

v01dXYZ replied:
The question you raise is actually quite important as actually this code is disabled with this current default config. I disabled it as I wanted to make it optimisation level dependent but I needed some input to calibrate it according to the level.

It is set to `1` which means the set of BBs that will be affected are `dist(Loop, BB) < 1` (the strict bound is on purpose). The `0` value is a special value for an `+infinity` bound. 

+1 for making `pushIVUsers` a method instead of functions.

I think I'll put it to `0` now as it will show performance regressions and allow discussing about that. I'll add some tests before though.

nikic commented on llvm/lib/Transforms/Utils/SimplifyIndVar.cpp:
```diff
@@ -854,16 +861,23 @@ static void pushIVUsers(
-    SimpleIVUsers.push_back(std::make_pair(UI, Def));
+    unsigned Counter =
+        L->contains(UI)
+            ? 0 // reset depth if we go back inside the loop.
```
Do we need to do that?

v01dXYZ replied:
No.

If we want to compute a overapproximation of the distance in the CFG `dist(Loop, BB)`, this helps to tighten the approximation to the real distance.

nikic commented on llvm/lib/Transforms/Utils/SimplifyIndVar.cpp:
```diff
@@ -854,16 +861,23 @@ static void pushIVUsers(
-    // Loops).
-    if (!L->contains(UI))
+    // Avoid adding Defs that SCEV expand to themselves, e.g. the LoopPhis
+    // of the outter loops.
```
```suggestion
    // of the outer loops.
```

nikic commented on llvm/lib/Transforms/Utils/SimplifyIndVar.cpp:
```diff
@@ -854,16 +861,23 @@ static void pushIVUsers(
+    unsigned Counter =
+        L->contains(UI)
+            ? 0 // reset depth if we go back inside the loop.
+            : OutOfLoopChainCounter + (UI->getParent() != Def->getParent());
```
Why is this a limit on the depth of the blocks rather than instructions? You can have a very long chain of instructions within a single block, or a very short one across blocks. It seems like that would be the more relevant quantity.

v01dXYZ replied:
At first, I implemented a instruction-wise depth but I changed that as I visualised the algorithm at the CFG level,  as accepting optimising basic blocks that are at given distance of the loop `dist(Loop, BB) < limit`. 

Also, I didn't like the idea to have partially optimised basic-blocks.

nikic commented on llvm/lib/Transforms/Utils/SimplifyIndVar.cpp:
```diff
@@ -854,16 +861,23 @@ static void pushIVUsers(
+
+    if (!MaxOutOfLoopChainCounter || Counter < MaxOutOfLoopChainCounter) {
+      SimpleIVUsers.push_back(std::make_tuple(UI, Def, Counter));
+    }
```
Remove braces.

nikic commented on llvm/lib/Transforms/Utils/SimplifyIndVar.cpp:
```diff
@@ -942,11 +956,11 @@ void SimplifyIndvar::simplifyUsers(PHINode *CurrIV, IVVisitor *V) {
       for (Use &U : UseInst->uses()) {
         Instruction *User = cast<Instruction>(U.getUser());
-        if (replaceIVUserWithLoopInvariant(User))
+        if (DT->dominates(L->getHeader(), User->getParent()) &&
```
Why do we need this condition? Aren't you already checking it when pushing uses?

v01dXYZ replied:
We checked `UseInst` is dominated but not its users.

We need this condition to "protect" the following call `replaceIVUserWithLoopInvariant` to ensure we don't feed it with a phi that is already invariant. It was not necessary before because the loop is LCSSA and any exits would be fenced with phis.

I could have misunderstood your point though.

nikic replied:
I see, thanks for the explanation.

nikic reviewed (commented):
Please add a test in llvm/test/Transforms/IndVarSimplify.

nikic commented:
No compile-time impact: https://llvm-compile-time-tracker.com/compare.php?from=e5a41f0afc152cc24b8fef3aa177ef53b2e77c43&to=96dd00646ea8346b9eb4efc0b5b5c78375e8189b&stat=instructions:u

Commit:
Add simple tests

Commit:
Default to enable full traversal (it shoud be changed before merge)

Commit:
[SimplifyIndVar] Push more users to worklist for simplifyUsers

Instead of considering only users that are inside the loop,
consider all the users with parent dominated by the loop header.

Commit:
typo

v01dXYZ commented:
> No compile-time impact: https://llvm-compile-time-tracker.com/compare.php?from=e5a41f0afc152cc24b8fef3aa177ef53b2e77c43&to=96dd00646ea8346b9eb4efc0b5b5c78375e8189b&stat=instructions:u

I changed the default config which disabled the traversal of users that are out of loop.

I expect performance regressions with reenabling full traversal.

Commit:
Update tests

Commit:
Update test for LoopDeletion/LoopVectorize

v01dXYZ commented:
I don't understand why the CI pipeline failed. Is there a way to relaunch it for linux without repushing the last commit ?

v01dXYZ commented:
Ping

nikic commented:
It looks like the updated patch miscompiles clang: http://llvm-compile-time-tracker.com/show_error.php?commit=9236a1de08a0f522a4c123b3818ea1dcb0361c3c There are crashes when building llvm-test-suite with stage2 clang.

Commit:
Fix eliminateTrunc

Out of loop, an IV user is loop invariant. Using L->isLoopInvariant is
no more valid to distinguish which operand is the loop-constant one from
the IV user.

nikic commented:
Can you please also add a test for your last fix?

We do get a working stage2 build now: https://llvm-compile-time-tracker.com/compare.php?from=3c50cbfda4fc3ad85349167132f7ed809ecc685a&to=8c47a1722b8b303e3262a77f992fea606a0a305c&stat=instructions:u

Commit:
pre-commit test eliminateTrunc miscompile

v01dXYZ commented:
I want to add another point to the ones mentioned by @efriedma-quic:

* the optimisations done in-loop are much more susceptible to improve the perf compared to out-of-loop. It's  more profitable compared to colder paths. It could be wiser to use a light version instead.
